File: internal/handler/auth/confirm_email_warning.go
package auth

import (
	"fmt"
	"html/template"
	"net/http"

	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
	"gitlab.com/gitlab-org/gitlab-redux/internal/util"
)

// ConfirmEmailWarning handles displaying a warning to users who need to confirm their email
type ConfirmEmailWarning struct {
	settingsService *service.SettingsService
	logger          *util.Logger
	email           string
}

// NewConfirmEmailWarning creates a new instance of ConfirmEmailWarning
func NewConfirmEmailWarning(
	settingsService *service.SettingsService,
	logger *util.Logger,
) *ConfirmEmailWarning {
	return &ConfirmEmailWarning{
		settingsService: settingsService,
		logger:          logger,
	}
}

// SetConfirmWarning sets a warning message if the user needs to confirm their email
func (c *ConfirmEmailWarning) SetConfirmWarning(ctx *gin.Context) {
	// Check if we should show the warning
	if !c.shouldShowConfirmWarning(ctx) {
		return
	}

	// Get the current user from the context
	currentUser, exists := ctx.Get("current_user")
	if !exists {
		return
	}
	user := currentUser.(*model.User)

	// Return if the user is already confirmed
	if user.Confirmed() {
		return
	}

	// Get the email to display
	email := c.getEmail(user)

	// Create the warning message
	message := c.formatConfirmWarningMessage(
		email,
		c.createResendLink(email),
		c.createUpdateLink(),
	)

	// Set the warning message in the flash
	ctx.Set("flash_warning", template.HTML(message))
}

// shouldShowConfirmWarning checks if we should show the confirm warning
func (c *ConfirmEmailWarning) shouldShowConfirmWarning(ctx *gin.Context) bool {
	// Check if it's an HTML request
	if ctx.GetHeader("Accept") != "text/html" {
		return false
	}

	// Check if it's a GET request
	if ctx.Request.Method != http.MethodGet {
		return false
	}

	// Check if email confirmation is set to soft
	return c.settingsService.EmailConfirmationSettingSoft()
}

// getEmail gets the email to display
func (c *ConfirmEmailWarning) getEmail(user *model.User) string {
	if c.email != "" {
		return c.email
	}

	if user.UnconfirmedEmail != "" {
		c.email = user.UnconfirmedEmail
	} else {
		c.email = user.Email
	}

	return c.email
}

// formatConfirmWarningMessage formats the confirm warning message
func (c *ConfirmEmailWarning) formatConfirmWarningMessage(email, resendLink, updateLink string) string {
	return fmt.Sprintf(
		"Please check your email (%s) to verify that you own this address and unlock the power of CI/CD. "+
			"Didn't receive it? %s. Wrong email address? %s.",
		template.HTMLEscapeString(email),
		resendLink,
		updateLink,
	)
}

// createResendLink creates a link to resend the confirmation email
func (c *ConfirmEmailWarning) createResendLink(email string) string {
	return fmt.Sprintf(
		`<a href="/users/confirmation?user[email]=%s" data-method="post">Resend it</a>`,
		template.HTMLEscapeString(email),
	)
}

// createUpdateLink creates a link to update the email address
func (c *ConfirmEmailWarning) createUpdateLink() string {
	return `<a href="/-/profile">Update it</a>`
}

File: rust_app/src/controllers/concerns/confirm_email_warning.rs
use actix_web::{HttpRequest, HttpResponse};
use std::sync::OnceLock;

pub struct Settings {
    pub email_confirmation_setting_soft: bool,
}

impl Settings {
    pub fn get() -> &'static Settings {
        static INSTANCE: OnceLock<Settings> = OnceLock::new();
        INSTANCE.get_or_init(|| Settings {
            email_confirmation_setting_soft: true,
        })
    }
}

pub struct User {
    pub email: String,
    pub unconfirmed_email: Option<String>,
    pub confirmed: bool,
}

pub trait ConfirmEmailWarning {
    fn show_confirm_warning(&self, req: &HttpRequest) -> bool {
        self.is_html_request(req)
            && req.method() == "GET"
            && Settings::get().email_confirmation_setting_soft
    }

    fn set_confirm_warning(&self, req: &HttpRequest) -> Option<String> {
        let user = self.get_current_user()?;
        if user.confirmed {
            return None;
        }

        let email = self.get_email(user);
        let warning_message = format!(
            "Please check your email ({}) to verify that you own this address and unlock the power of CI/CD. \
            Didn't receive it? <a href='/users/confirmation?email={}'>Resend it</a>. \
            Wrong email address? <a href='/users/settings/profile'>Update it</a>.",
            self.escape_html(&email),
            self.escape_html(&email)
        );

        Some(warning_message)
    }

    fn get_email(&self, user: &User) -> String {
        user.unconfirmed_email
            .clone()
            .unwrap_or_else(|| user.email.clone())
    }

    fn escape_html(&self, text: &str) -> String {
        text.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;")
    }

    // Required methods to be implemented by concrete types
    fn is_html_request(&self, req: &HttpRequest) -> bool;
    fn get_current_user(&self) -> Option<&User>;
}
